File: src/main/java/com/lcaohoanq/nocket/domain/auth/AuthPort.java
package com.lcaohoanq.nocket.domain.auth;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.lcaohoanq.nocket.constant.Regex;
import com.lcaohoanq.nocket.enums.Country;
import com.lcaohoanq.nocket.enums.Currency;
import com.lcaohoanq.nocket.enums.Gender;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Pattern;

public interface AuthPort {

    record AccountRegisterDTO(
        @JsonProperty("name")
        @NotBlank(message = "Name is required") String name,

        @JsonProperty("email")
        @Email(message = "Email is invalid") String email,

        @JsonProperty("phone_number")
        @Pattern(regexp = Regex.PHONE_NUMBER_REGEX, message = "Phone number is invalid")
        @NotBlank(message = "Phone number is required") String phoneNumber,

        @JsonProperty("gender") Gender gender,

        @JsonProperty("password")
        @Pattern(regexp = Regex.PASSWORD_REGEX, message = "Password must contain at least one uppercase letter, one lowercase letter, one number, and one special character")
        @NotBlank(message = "Password is required") String password,

        @JsonProperty("confirm_password")
        @NotBlank(message = "Confirm password is required") String confirmPassword,

        @JsonProperty("date_of_birth")
        String dateOfBirth,

        @JsonProperty("preferred_language")
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        Country preferredLanguage,

        @JsonProperty("preferred_currency")
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        Currency preferredCurrency

    ){}

    record UpdatePasswordDTO(
        @JsonProperty("email")
        @Email(message = "Email must be a valid email")
        @NotBlank(message = "Email must not be blank")
        String email,

        @JsonProperty("new_password")
        @Pattern(regexp = Regex.PASSWORD_REGEX, message = "Password must contain at least one uppercase letter, one lowercase letter, one number, and one special character")
        @NotBlank(message = "New password must not be blank")
        String newPassword
    ) {}

    record UserLoginDTO(
        @JsonProperty("email")
        @Email(message = "Email is not valid")
        @NotBlank(message = "Email is required")
        String email,

        @JsonProperty("password")
        @NotBlank(message = "Password is required")
        String password
    ) {}

    record VerifyUserDTO(
        @JsonProperty("email")
        @NotBlank(message = "Email is required") String email,

        @JsonProperty("otp")
        @NotBlank(message = "OTP is required") String otp
    ) {}
    
}

File: src/main/java/com/lcaohoanq/nocket/domain/auth/AuthPort.kt
package com.lcaohoanq.nocket.domain.auth

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonProperty
import com.lcaohoanq.nocket.constant.Regex
import com.lcaohoanq.nocket.enums.Country
import com.lcaohoanq.nocket.enums.Currency
import com.lcaohoanq.nocket.enums.Gender
import jakarta.validation.constraints.Email
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.Pattern

interface AuthPort {

    data class AccountRegisterDTO(
        @JsonProperty("name")
        @NotBlank(message = "Name is required") val name: String,

        @JsonProperty("email")
        @Email(message = "Email is invalid") val email: String,

        @JsonProperty("phone_number")
        @Pattern(regexp = Regex.PHONE_NUMBER_REGEX, message = "Phone number is invalid")
        @NotBlank(message = "Phone number is required") val phoneNumber: String,

        @JsonProperty("gender") val gender: Gender,

        @JsonProperty("password")
        @Pattern(
            regexp = Regex.PASSWORD_REGEX,
            message = "Password must contain at least one uppercase letter, one lowercase letter, one number, and one special character"
        )
        @NotBlank(message = "Password is required") val password: String,

        @JsonProperty("confirm_password")
        @NotBlank(message = "Confirm password is required") val confirmPassword: String,

        @JsonProperty("date_of_birth")
        val dateOfBirth: String,

        @JsonProperty("preferred_language")
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        val preferredLanguage: Country,

        @JsonProperty("preferred_currency")
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        val preferredCurrency: Currency
    )

    data class UpdatePasswordDTO(
        @JsonProperty("email")
        @Email(message = "Email must be a valid email")
        @NotBlank(message = "Email must not be blank")
        val email: String,

        @JsonProperty("new_password")
        @Pattern(
            regexp = Regex.PASSWORD_REGEX,
            message = "Password must contain at least one uppercase letter, one lowercase letter, one number, and one special character"
        )
        @NotBlank(message = "New password must not be blank")
        val newPassword: String
    )

    data class UserLoginDTO(
        @JsonProperty("email")
        @Email(message = "Email is not valid")
        @NotBlank(message = "Email is required")
        val email: String,
        @JsonProperty("password")
        @NotBlank(message = "Password is required")
        val password: String
    )

    data class VerifyUserDTO(
        @JsonProperty("email")
        @NotBlank(message = "Email is required") val email: String,
        @JsonProperty("otp")
        @NotBlank(message = "OTP is required") val otp: String
    )

}